### DISCUSSION TITLE
different ways and methods to essentially do the same, across components

### DISCUSSION THREAD

**Discussion by vincentbriglia** [chakra-ui/chakra-ui#3692] [General]
Hi, before taking a crack at what would be a massively breaking change, I wanted to have a discussion around the discrepancies in uniformity of a handful of components, and specifically in the way you action them. 

Many components that in my opinion should have the same kind of initialization and capabilities, don't. 

the components i'm talking about are `Menu`, `Tooltip`, `Popover`, `Drawer`, `Modal` and `AlertDialog` (i might be missing one)

f.ex `Menu` has an embedded `MenuButton`, whereas `Tooltip` wraps anything you want to add a tooltip to (which is fine, i totally get that) but then `Popover` also has an embedded `PopoverTrigger` but when we then look at `Drawer`, `Modal` and `AlertDialog` they all have external Triggers.

Most, not all of these components have render props, to access the internal state, which is handy if you want to show whether something is active, disabled or open ... but some don't, f.ex the `Drawer` doesn't expose `isOpen` as a render prop, because you need to manage the state outside, whereas the `Menu` clearly does. 

I'd like to see (bar Tooltip) that all these components have an overlap in their API. I would like to see all of them have the ability to be Controlled/Uncontrolled and for all of them to then also have render props at the top level. 

is this something that's an oversight? or deliberate, either way it would be great to have a discussion.

## Menu

inner trigger, `Menu` has render props

```
<Menu>
  <MenuButton as={Button} rightIcon={<ChevronDownIcon />}>
    Actions
  </MenuButton>
  <MenuList>
    <MenuItem>Download</MenuItem>
    <MenuItem>Create a Copy</MenuItem>
    <MenuItem>Mark as Draft</MenuItem>
    <MenuItem>Delete</MenuItem>
    <MenuItem>Attend a Workshop</MenuItem>
  </MenuList>
</Menu>
```

wrapper, no render props, fine as it is, API makes sense.

## Tooltip
```
<Tooltip label="I close on click">
  <Button>Close on Click - true(default)</Button>
</Tooltip>
```

## Popover

inner trigger, but can also have controlled nature 👍🏼  , `Popover` has render props

```
<Popover>
  <PopoverTrigger>
    <Button>Trigger</Button>
  </PopoverTrigger>
  <PopoverContent>
    <PopoverArrow />
    <PopoverCloseButton />
    <PopoverHeader>Confirmation!</PopoverHeader>
    <PopoverBody>Are you sure you want to have that milkshake?</PopoverBody>
  </PopoverContent>
</Popover>
```

## Drawer

no internal trigger, manage yourself, no render props

```
      <Button ref={btnRef} colorScheme="teal" onClick={onOpen}>
        Open
      </Button>
      <Drawer
        isOpen={isOpen}
        placement="right"
        onClose={onClose}
        finalFocusRef={btnRef}
      >
        <DrawerOverlay>
         ...
```

## AlertDialog

no internal trigger, manage yourself, no render props

```
      <Button colorScheme="red" onClick={() => setIsOpen(true)}>
        Delete Customer
      </Button>

      <AlertDialog
        isOpen={isOpen}
        leastDestructiveRef={cancelRef}
        onClose={onClose}
      >
        <AlertDialogOverlay>
```

## Modal

no internal trigger, manage yourself, no render props

```
      <Button onClick={onOpen}>Open Modal</Button>

      <Modal isOpen={isOpen} onClose={onClose}>
        <ModalOverlay />
```

**Comment by chasinhues** [comment]
If possible, unifying the way in which these API's are used makes sense to me. 

I just ran into an issue with `Menu` specifically. I want to use it, but I want to control how it is opened. I want to implement an autocomplete component where typing an input summons the menu with dynamically loaded items. Seems I can't really do that at the moment with the current API. 

Could all of these dialog components be using the disclosure pattern that `Modal` is using perhaps?
